lib/device.c

```diff
@@ -237,6 +271,49 @@ static SOCKET server_connect(const char *host, unsigned short nport)
 	return sock;
 }
 
+static int server_greet(struct sync_device *d)
+{
+	char greet[128];
+
+	if (sockio_send(d, CLIENT_GREET, strlen(CLIENT_GREET)) ||
+	    sockio_recv(d, greet, strlen(SERVER_GREET)) ||
+	    strncmp(SERVER_GREET, greet, strlen(SERVER_GREET))) {
+		sockio_close(d);
+		return -1;
+	}
+
+	return 0;
+}
+
+/* Setting a new cb+ctxt establishes a new abstract connection represented by ctxt.
+ * How the connection communicates is implemented by the .recv()/.send() members.
+ * How the connection ends and cleans itself up is implemented by .close().
+ * ctxt must already be allocated, connected, and ready to be used with the provided methods.
+ * The device takes ownership of ctxt, to be cleaned up by the .close() method to disconnect.
+ *
+ * Returns 0 on success, -1 on error, which may occur since this performs the
+ * initial Rocket handshake.  Even in error ctxt will be closed.
+ *
+ * This is the low-level communications api, only use this if the sync_tcp_connect() style
+ * helper is insufficient for your needs.  They are mutually exclusive.
+ */
+int sync_set_sockio_cb(struct sync_device *d, struct sync_sockio_cb *cb, void *ctxt)
+{
+	assert(d);
+	assert(ctxt);
+	assert(cb->send);
+	assert(cb->recv);
+	assert(cb->close);
```

Comment:
I removed most of the asserts, but just kept this batch since they police a significant programmer error in a way that fails fast near the error's origin vs. when something tries to blindly use send/recv/close - and everything else wants to be able to assume these are either all set or sockio_ctxt is NULL.   And the NULL sockio_ctxt is used to indicate there's no valid sockio established, installing methods with a NULL ctxt is just broken unexpected usage.  Only the library sets a NULL ctxt either as the initial state, or in sockio_close().

I'm accustomed to gratuitous asserts as a sort of defensive programming that also serves to document API expectations/assumptions.  But I think the current state is probably a decent compromise?  If you insist on removing more, it's your project, I'll remove them :)